File: Java/compoundlab/src/main/java/com/bbd/abstractFactoryPattern/RedheadDuck.java
package com.bbd.abstractFactoryPattern;

public class RedheadDuck implements IQuackable
{
	IQuackObservable _observable;
	private String _name = "Red Head Duck";

	RedheadDuck()
	{
		this._observable = new Observable(this);
	}

	RedheadDuck(IQuackObservable observable)
	{
		this._observable = observable;
	}

	public String quack()
	{
		notifyObservers();
		return ("Red head duck quacks");
	}

	public void registerObserver(IObserver observer)
	{
		this._observable.registerObserver(observer);
	}

	public String notifyObservers()
	{
		return (this._observable.notifyObservers());
	}

	@Override
	public String getName() { return  this._name;}
}

File: Scala/compoundlab/src/main/scala/RedheadDuck.scala


class RedheadDuck(observable : Option[IQuackableObservable] = None) extends IQuackable {

	private val name = "Red head Duck"
	private val _observable : IQuackableObservable = observable getOrElse new Observable(this)


	override def registerObserver(observer : IObserver) : Unit = {
		this._observable.registerObserver(observer)
	}

	override def notifyObservers: String = {
		this._observable.notifyObservers
	}

	override def quack : String = {
		notifyObservers
		"Red Head Duck quacked"
	}

	override def getName: String = {
		this.name
	}
}
